### PR TITLE
WT-7853 Clear the update datastore flag if reconciliation fails

### PR DESCRIPTION
**Pull request by luke-pearson**

### CODE REVIEW

File: src/include/btmem.h [hunk 1]
```diff
@@ -1131,8 +1131,13 @@ struct __wt_update {
      */
     volatile uint8_t prepare_state; /* prepare state */
 
+/*
+ * WT_UPDATE_DS indicates that an update is being written or has been written to the data store
+ * during a round of reconciliation. This update could be cleared if the reconciliation fails. It is
```
Comment (quchenhao) [inline src/include/btmem.h:1136]: This flag could be cleared?

File: src/reconcile/rec_write.c [hunk 2]
```diff
@@ -668,6 +671,26 @@ __rec_init(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t flags, WT_SALVAGE_COO
     return (ret);
 }
 
+/*
+ * __rec_clear_datastore_update_list --
+ *     Clear the update cache stored on the reconciliation structure, if called from the error path
+ *     then clear the update datastore flag.
+ */
+static void
+__rec_clear_datastore_update_list(WT_SESSION_IMPL *session, WT_RECONCILE *r, bool error)
```
Comment (quchenhao) [inline src/reconcile/rec_write.c:680]: Are you sure we call this before we release the eviction's exclusive lock?

File: test/suite/test_hs_evict_race01.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
+#!/usr/bin/env python
```
Comment (quchenhao) [inline test/suite/test_hs_evict_race01.py:1]: I think this should be a checkpoint test? evict_race is not clear as it doesn't say what it races with. I prefer just eviction test or checkpoint test.

Comment (quchenhao) [reply test/suite/test_hs_evict_race01.py:1]: I looked the code. It's ok.

File: test/suite/test_hs_evict_race01.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+        # Insert a value at timetamp 5
+        self.session.begin_transaction()
+        cursor[1] = self.value2
+        self.session.commit_transaction('commit_timestamp=' + self.timestamp_str(5))
+
+        cursor.close()
+
+        # Create a thread.
+        ooo_thread = threading.Thread(target=self.out_of_order_update_and_evict)
+        self.session.breakpoint()
```
Comment (quchenhao) [inline test/suite/test_hs_evict_race01.py:82]: Is this still needed?

File: src/reconcile/rec_write.c (hunk 2 above)
Comment (luke-pearson) [reply src/reconcile/rec_write.c:680]: Yes this is called before eviction's lock is released, additionally in the error case it is called before the reconciliation page lock is released.

File: src/include/btmem.h (hunk 1 above)
Comment (luke-pearson) [reply src/include/btmem.h]: Fixed

File: test/suite/test_hs_evict_race01.py (hunk 3 above)
Comment (luke-pearson) [reply test/suite/test_hs_evict_race01.py:1]: Could be a test_bug? I like more explicit names though. test_hs_checkpoint_eviction_race?

**Review by quchenhao — APPROVED** [review]
lgtm

File: src/reconcile/rec_visibility.c [hunk 5]
```diff
@@ -582,10 +613,10 @@ __wt_rec_upd_select(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins, v
          * for the data store. Subsequent reconciliations should know that they can select this
          * update regardless of visibility.
          */
-        if (upd_select->upd != NULL)
-            F_SET(upd_select->upd, WT_UPDATE_DS);
-        if (tombstone != NULL)
-            F_SET(tombstone, WT_UPDATE_DS);
+        if (upd_select->upd != NULL && !F_ISSET(upd_select->upd, WT_UPDATE_DS))
+            __rec_add_datastore_update(session, r, upd_select->upd);
```
Comment (bvpvamsikrishna) [inline src/reconcile/rec_visibility.c:617]: error handling if this returns an error, in case "__wt_calloc_one" failed.

Comment (luke-pearson) [reply src/reconcile/rec_visibility.c]: 01c2d22

**Review by bvpvamsikrishna — APPROVED** [review]
lgtm

**Pull request merged**